src/rust/engine/fs/fs_util/src/main.rs

```diff
@@ -100,56 +131,111 @@ fn execute(top_match: clap::ArgMatches) -> Result<(), String> {
   }
 }
 
-fn save_file(store: &Store, path: &Path) -> Result<(Fingerprint, usize), String> {
-  let mut buf = Vec::new();
-  File::open(path)
-    .and_then(|mut f| f.read_to_end(&mut buf))
-    .map_err(|e| {
-      format!("Error reading file {:?}: {}", path, e.description())
-    })?;
-  Ok((store.store_file_bytes(&buf)?, buf.len()))
+fn make_posix_fs(root: &Path) -> fs::PosixFS {
+  fs::PosixFS::new(&root, vec![]).unwrap()
 }
 
-fn save_directory(store: &Store, root: &Path) -> Result<(Fingerprint, usize), String> {
-  let mut directory = bazel_protos::remote_execution::Directory::new();
-  for entry in fs::PosixFS::scandir_sync(root.to_path_buf(), fs::Dir(root.to_path_buf()))
-    .map_err(|e| {
-      format!("Error listing directory {:?}: {}", root, e.description())
-    })?
-  {
-    match entry {
-      fs::Stat::Dir(fs::Dir(path)) => {
-        let (fingerprint, proto_size_bytes) = save_directory(store, &path)?;
-        directory.mut_directories().push({
-          let mut dir_node = bazel_protos::remote_execution::DirectoryNode::new();
-          dir_node.set_name(file_name_as_utf8(&path)?);
-          dir_node.set_digest(fingerprint_to_digest(&fingerprint, proto_size_bytes as i64));
-          dir_node
-        })
-      }
-      fs::Stat::File(fs::File {
-                       path,
-                       is_executable,
-                     }) => {
-        let (fingerprint, size_bytes) = save_file(store, &path)?;
-        directory.mut_files().push({
-          let mut file_node = bazel_protos::remote_execution::FileNode::new();
-          file_node.set_name(file_name_as_utf8(&path)?);
-          file_node.set_digest(fingerprint_to_digest(&fingerprint, size_bytes as i64));
-          let mut absolute_path = root.to_path_buf();
-          absolute_path.push(path.file_name().ok_or_else(|| {
-            format!("{:?} did not have a file_name", path)
-          })?);
-          file_node.set_is_executable(is_executable);
-          file_node
-        });
-      }
-      fs::Stat::Link(fs::Link(path)) => {
-        return Err(format!("Don't yet know how to handle symlinks: {:?}", path))
-      }
-    }
-  }
-  store.record_directory(&directory)
+fn save_file(
+  store: Arc<Store>,
+  posix_fs: &fs::PosixFS,
+  file: fs::File,
+) -> BoxFuture<(Fingerprint, usize), String> {
+  posix_fs
+    .read_file(&file)
+    .map_err(move |err| {
+      format!("Error reading file {:?}: {}", file, err.description())
+    })
+    .and_then(move |content| {
+      Ok((
+        store.store_file_bytes(&content.content)?,
+        content.content.len(),
+      ))
+    })
+    .to_boxed()
+}
+
+fn save_directory(
+  store: Arc<Store>,
+  posix_fs: Arc<fs::PosixFS>,
+  base_dir: Arc<fs::Dir>,
+) -> BoxFuture<(Fingerprint, usize), String> {
+  let store_copy = store.clone();
```

Comment:
Hm, not sure if I'm reading this right, but I don't think that this alias of store is used... `store` moves into the closure, and is cloned there for the children.

Reply:
store_copy is used in the very last `and_then`'s closure, for store_copy.record_directory(&directory)

Would it be more clear if store_copy were moved into the first `and_then`, and store was used in the last one? Probably a little... I'll do that.